### DISCUSSION TITLE
leaf + wintun 如何实现tun模式

### DISCUSSION THREAD

**Discussion by stoneport** [eycorsican/leaf#343] [General]
想实现一个Windows下Trojan TUN模式客户端，考虑使用leaf + wintun做二次开发。目前能做到windows下启动leaf，也能安装wintun虚拟网卡，但就是不知道如何将二者关联起来。

leaf的客户端配置文件：

```
[General]
loglevel = trace
dns-server = 8.8.8.8, 1.1.1.1
socks-interface = 127.0.0.1
socks-port = 1086

# windows下如何关联tun?
# tun-fd = ?
tun = auto

[Proxy]
Direct = direct
Proxy = trojan, myip, 443, password=mypassword, sni=mydomain

[Rule]
FINAL, Proxy
```

粗略学习了一下[YtFlow/Maple](https://github.com/YtFlow/Maple)项目，发现该项目仅支持windows 10以后的UWP。虽然省掉了第三方的TUN/TAP驱动，不过对Windows版本的要求比较高。

请指教，谢谢！

**Comment by eycorsican** [comment]
主要看怎么用Rust调用wintun的API，主要实现函数就两个，读IP包和写IP包，做好了再考虑放进leaf中。不管是换tun还是netstack，leaf相关的改动最多就一百多行: https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L160-L293，最直观的方法是用unbounded channel传IP包。

**Reply by stoneport** [reply]
非常感谢指出关键的方向。我以前没搞过rust开发，先尝试一下看看能不能搞懂这部分rust代码。

我的程序使用go开发，目前卡在这个tun2socks上了。整理了三个方案。

#### 方案一：使用外部的[xjasonlyu/tun2socks](https://github.com/xjasonlyu/tun2socks)

leaf负责启动本地的socks服务，然后调用外部的tun2socks独立命令行转发到wintun虚拟网卡。
经验证简单可行，不过理论上I/O次数过多，通信效率相对较差。

#### ~~方案二：研究wintun的go API，创建`fd-based endpoint`，迎合leaf的配置文件。~~
~~该方案的好处是leaf不做修改，保持原装。~~

更新：这个方案不现实。FD (File Descriptor)只有在Unix家族系统里才能用得上，Windows对于进程之间的数据读写的方式不同。
参考： [Lacking Rhoticity: Passing FDs/handles between processes on Unix and Windows -- a comparison](http://lackingrhoticity.blogspot.com/2015/05/passing-fds-handles-between-processes.html)

#### 方案三：修改leaf中tun2socks或者配置接口部分的rust代码，迎合wintun的接口

[eycorsican](https://github.com/eycorsican)大神指出的方向。该方案能用上leaf内部的tun2socks功能，理论上这种方式运行效率最高。

对于rust的外行，存在一定的学习门槛。

**Reply by rnatjaco** [reply]
> 主要看怎么用Rust调用wintun的API，主要实现函数就两个，读IP包和写IP包，做好了再考虑放进leaf中。不管是换tun还是netstack，leaf相关的改动最多就一百多行: [https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L160-L293，最直观的方法是用unbounded](https://github.com/eycorsican/leaf/blob/master/leaf/src/proxy/tun/inbound.rs#L160-L293%EF%BC%8C%E6%9C%80%E7%9B%B4%E8%A7%82%E7%9A%84%E6%96%B9%E6%B3%95%E6%98%AF%E7%94%A8unbounded) channel传IP包。

我正在尝试做这件事，目前已经替换了支持 windows 的 tun 库，现在可能因为 post_tun_creation_setup 和 post_tun_completion_setup 导致回环（现象是会一直刷 outbound 的日志），我怀疑是路由导致的，有什么思路吗？

**Reply by eycorsican** [reply]
@rnatjaco windows 上不需要 post_tun_creation_setup 这些，需要把 tun interface 的 metric 设置为 0，然后最好支持一下设置 tun interface 的 dns 服务器，这些 rust-tun 都已经支持，然后创建 tun 前还要设置一下 OUTBOUND_INTERFACE 为原默认网卡的 IP，比如可以这样：

```
#[cfg(windows)]
{
    let s = std::net::UdpSocket::bind("0.0.0.0:0")?;
    s.connect("1.1.1.1:53")?;
    std::env::set_var("OUTBOUND_INTERFACE", s.local_addr()?.ip().to_string());
}
```

**Reply by eycorsican** [reply]
rust-tun 在创建 tun 后已经加了一条默认路由，windows 选取网关一个是看 local address，一个是看 metric (interface metric + route metric)，设置了 OUTBOUND_INTERFACE 后因为 local address 是原网卡的 IP 所以 outbound 出去的流量会走原网关，其它进程 local address 一般没有绑定某个 IP，就由 metric 决定走哪个网关，metric 越低优先级越高，所以走 tun，被 leaf 接收，再经由 leaf 的 outbound 从原网关出去。

**Reply by lanthora** [reply]
@eycorsican 已发 PR，目前测试正常

> 设置 tun interface 的 dns 服务器

会影响到哪些功能？目前测试代理直连和阻断都正常

**Reply by eycorsican** [reply]
@lanthora 如果 tun 的 adapter 上没 dns，就会用原默认 adapter 上配置的 dns，而 dhcp 分配的 dns 很多都是网关上的 dns，这一般会导致 dns 请求绕过 tun 直接走原默认网关。metric 如果不设置为 0，估计也有概率 tun 的 metric 比原默认网关高，导致没流量进入 tun

**Comment by stoneport** [comment]
期待[eycorsican/leaf](https://github.com/eycorsican/leaf)能加入对wintun的支持。

收集了几个wintun rust相关的文档：
- [nulldotblack/wintun: Rust bindings to the wintun c library: https://www.wintun.net/](https://github.com/nulldotblack/wintun)
- [wintun - crates.io: Rust Package Registry](https://crates.io/crates/wintun)
- [xutianyi1999/SimpleWintun: Wintun rust API library](https://github.com/xutianyi1999/SimpleWintun)

具体的处理过程也许可以借鉴一下: [xjasonlyu/tun2socks](https://github.com/xjasonlyu/tun2socks)

**Comment by xSILENCEx** [comment]
@stoneport 你好，请问有进展吗？

**Reply by stoneport** [reply]
没搞定，后来我用Go直接调用clash premium外部程序了
